Vue 3 vueschool links

rahaug opened the pull request:
This PR adds Vue School links to the documentation to supplement the documentation with free training videos. 

The lessons are up-to-date, V3 versions of previously approved Vue 2 lessons that were added to the documentation in [PR 2379](https://github.com/vuejs/vuejs.org/pull/2379). As requested in [PR 311](https://github.com/vuejs/docs-next/pull/311)

[Screenshots of the placements can be seen here](https://imgur.com/a/1G4UGNo)

**Lessons**
[Class Binding](https://vueschool.io/lessons/dynamic-css-classes-with-vue-3)
[Event Handling](https://vueschool.io/lessons/user-events-in-vue-3)
[Computed Properties](https://vueschool.io/lessons/computed-properties-in-vue-3)
[Conditional Rendering](https://vueschool.io/lessons/conditional-rendering-in-vue-3)
[List Rendering](https://vueschool.io/lessons/list-rendering-in-vue-3)
[Methods](https://vueschool.io/lessons/methods-in-vue-3)
[Form Input](https://vueschool.io/lessons/user-inputs-vue-devtools-in-vue-3)
[Vue Devtools](https://vueschool.io/lessons/using-vue-dev-tools-with-vuejs-3)

bencodezen commented on src/guide/events.md:
```diff
@@ -1,5 +1,7 @@
 # Event Handling
 
+<VideoLesson href="https://vueschool.io/lessons/user-events-in-vue-3?friend=vuejs" title="Learn how to handle events on Vue School">Learn how to handle events in a free Vue School lesson</VideoLesson>
```
The lesson doesn't start in on the event material till halfway through. Not sure this is a good way to introduce people to a concept.

bencodezen commented on src/guide/list.md:
```diff
@@ -1,5 +1,7 @@
 # List Rendering
 
+<VideoLesson href="https://vueschool.io/lessons/list-rendering-in-vue-3?friend=vuejs" title="Learn how to render lists on Vue School">Learn how to render list with a free Vue School lesson</VideoLesson>
```
This lesson seemed a little light compared to the CSS class one which goes over the different ways you can bind to a dynamic class. Other scenarios missing include: how looping objects work, pulling out indices, and the recommendation for adding keys to each element to help Vue track of nodes.

bencodezen commented on src/guide/forms.md:
```diff
@@ -1,5 +1,7 @@
 # Form Input Bindings
 
+<VideoLesson href="https://vueschool.io/lessons/user-inputs-vue-devtools-in-vue-3?friend=vuejs" title="Learn how to handle form inputs on Vue School">Learn how to work with form inputs with a free Vue School lesson</VideoLesson>
```
For this video, as a new student I would expect a much more in-depth covering of form inputs across different input items, along with a more in-depth explanation of what v-model is. This video shows that v-model enables two-way binding, but then spends the rest of the video talking about DevTools. This seems mismatched for the docs.

bencodezen commented on src/guide/data-methods.md:
```diff
@@ -1,5 +1,7 @@
 # Data Properties and Methods
 
+<VideoLesson href="https://vueschool.io/lessons/methods-in-vue-3?friend=vuejs" title="Learn how to use methods on Vue School">Learn how to work with data and methods with a free Vue School lesson</VideoLesson>
```
Because this section covers both data and methods, it seems misleading to have this lesson here at the top of the doc.

bencodezen reviewed (changes requested):
Hey @rahaug. Thanks for the PR! I've left a few comments regarding some of the lesson submitted.

As a whole, I have a general concern about the lessons generally requiring the user to have knowledge established in previous lessons. 

For example, "Remember how we did XYZ in the last lesson?" While I'm able to figure out what's going on due to my experience, this seems like it could be confusing for new students trying to learn a concept when they won't have prior knowledge of what's going on.

Because the videos are placed so high on the individual docs page, we need to have the lessons provide a more holistic depiction of concepts since there's no guarantee that they will return back to the docs.

Let me know if you have any questions!

rahaug commented:
Hi @bencodezen, thanks for the detailed review and helpful feedback!

I appreciate that you take the time to thoroughly review our content and help us create better content for the community and the documentation.

We will work on the feedback and let you know.

rahaug commented:
Hi @bencodezen,

Thanks again for your time and contribution to the course. We have updated and re-recorded the lessons according to your feedback here and in the private messages.

Can you please review the changes and merge if everything looks good?

rahaug commented:
I've updated the PR and linked the video on [how to install Vue Devtools](https://vueschool.io/lessons/using-vue-dev-tools-with-vuejs-3), which also cover how to use devtools with local .HTML files as we've extracted that to an individual lesson.

bencodezen commented:
Hey @skirtles-code! Could I get your eyes on the new videos to see if I missed anything?

skirtles-code reviewed (commented):
Would it be possible to get closed captions on the videos?

I've made a few observations about individual videos below. I don't have a good sense of how difficult it is to make minor changes to these videos so I've reviewed them just as I would for text documentation.

rahaug commented:
Thanks for reviewing the lessons @skirtles-code  and good job with catching some things we've missed! We're happy to work with you to improve the lessons.

Great suggestion on the captions! We have captions on all of our videos. Captions for these lessons are in production and will be available next week :)

The video production is a bit complex, and it's not always straightforward to ımplement changes. Ben has been involved in the review process with the text versions, and we'd love to include you in the text-based reviews if you're interested before we record other lessons for the docs.

We can patch in some of the suggestions. However, I do suggest merging now while we implement the changes, as the lessons are valuable to the docs users at this point.

How does this sound? :) 

@danielkellyio can you please address the individual lesson's questions below?

danielkellyio commented on src/guide/forms.md:
```diff
@@ -1,5 +1,7 @@
 # Form Input Bindings
 
+<VideoLesson href="https://vueschool.io/lessons/user-inputs-vue-devtools-in-vue-3?friend=vuejs" title="Learn how to handle form inputs on Vue School">Learn how to work with form inputs with a free Vue School lesson</VideoLesson>
```
While it may be misleading, at this point, users don't know about components yet. So instead of diving into that rabbit hole, our intention is to excite the viewer

skirtles-code replied:
I think it is more likely to frustrate them when they waste hours trying to find documentation for a feature that doesn't exist.

danielkellyio replied:
agreed. I can cut out that sentence.

danielkellyio commented on src/guide/class-and-style.md:
```diff
@@ -3,6 +3,7 @@
 A common need for data binding is manipulating an element's class list and its inline styles. Since they are both attributes, we can use `v-bind` to handle them: we only need to calculate a final string with our expressions. However, meddling with string concatenation is annoying and error-prone. For this reason, Vue provides special enhancements when `v-bind` is used with `class` and `style`. In ...
 
 ## Binding HTML Classes
+<VideoLesson href="https://vueschool.io/lessons/dynamic-css-classes-with-vue-3?friend=vuejs" title="Free Vue.js Dynamic Classes Lesson">Watch a free video lesson on Vue School</VideoLesson>
```
At 1:05, when demoing the object syntax, what do you think about adding in the following audio? 
"The conditional provided as the value can be anything truthy or falsey. It can be an data property as shown here but it could also be a call to a method, a computed prop, or a ternary statement." Thanks!

skirtles-code replied:
I'm not sure how that would help.

At around the 2:00 mark some code is added like this:

```
:class="[item.purchased ? 'strikeout' : '']"
```

The array is not required. The 'correct' way to write that conditional would be:

```
:class="item.purchased ? 'strikeout' : ''"
```

The role of the array is purely as a combinator, allowing multiple strings and objects to be combined. The video presents arrays as though they're a required part of adding a conditional class via an expression but the two are largely orthogonal.

danielkellyio replied:
I see what you're saying. Thinking on this.

danielkellyio replied:
At 00:23 I've replaced "Vue has 2 ways for adding dynamic classes" 
with "Vue has 2 ways for adding dynamic classes in addition to the normal attribute binding. Those 2 ways are the array and object syntax."
Also around 2:46 I right after demonstrating the 
```
:class="[item.purchased ? 'strikeout' : '']"
```
I've added "It’s worth noting that I’m only using the array syntax here in order quickly move to multiple expressions in just a minute. If you only need one expression you can omit the array brackets all together and it works just like any other attribute binding."

skirtles-code replied:
I think this explanation is still problematic.

At about 1:49 it says it's going to demonstrate the array syntax. From then until 2:45, almost everything that gets said in the narration is about using a ternary expression, not the array syntax. e.g. 'It [the array syntax] allows adding multiple classes to one condition'. Explaining the use of ternaries is really valuable but I don't think it should presented as part of the array syntax. The array syntax itself isn't really explained during that section.

I think it would be better to discuss the ternary approach before showing objects or arrays. The array syntax can then be presented just as a way to combine multiple expressions.

danielkellyio commented on src/guide/conditional.md:
```diff
@@ -1,5 +1,7 @@
 # Conditional Rendering
 
+<VideoLesson href="https://vueschool.io/lessons/conditional-rendering-in-vue-3?friend=vuejs" title="Learn how conditional rendering works with Vue School">Learn how conditional rendering works with a free lesson on Vue School</VideoLesson>
```
Nice catch! Though definitely a typo, I don't think it's a blocker to understanding the conditional rendering. wdyt? Thanks!

skirtles-code replied:
Agreed.

danielkellyio commented on src/guide/computed.md:
```diff
@@ -2,6 +2,8 @@
 
 ## Computed Properties
 
+<VideoLesson href="https://vueschool.io/lessons/computed-properties-in-vue-3?friend=vuejs" title="Learn how computed properties work with Vue School">Learn how computed properties work with a free lesson on Vue School</VideoLesson>
```
Nice catch! this is definitely an error. I'll have the audio patched asap with: 
"to avoid that we can use the spread operator to make a clone..."

skirtles-code replied:
Sounds good.

danielkellyio commented:
@skirtles-code thanks for taking the time to look over this and for the feedback! I've addressed the individual issues above. Let me know if you have any further questions/comments? Thanks!

danielkellyio commented:
@skirtles-code @rahaug the audio patches mentioned in the individual comments above have been uploaded and are live on the site. Thanks!

danielkellyio commented:
@skirtles-code I have created an alternate PR with all the lessons discussed that have no more blockers. what do you think of merging without the dynamic css classes video for now?
https://github.com/vuejs/docs-next/pull/1072
Thanks!

cc @hootlex @rahaug

bencodezen reviewed (approved):
Thanks for the hard work on this @danielkellyio!
